### PR TITLE
feat: Avoid unzipping of real device .ipa bundles

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
We can install iOS apps on real devices without unzipping them. This should allow to save time on session startup as we don't need to unzip anymore. Also the installation itself should perform better as the size of a compressed .ipa file is lower than the size of an uncompressed one, and we are always limited by the (USB2.0) transfer speed.

### CODE REVIEW

**Comment by KazuCocoa** [comment]
Will this drops `ios-deploy` as well? I may need to find better test app which worked with `ios-deploy` but not in current installation method in the past if this implementation can cover such a case

**Comment by mykola-mokhnach** [comment]
> Will this drops `ios-deploy` as well? I may need to find better test app which worked with `ios-deploy` but not in current installation method in the past if this implementation can cover such a case. It will take a bit time to prepare test app (maybe the test app should not be a small sample app)

Yes it does. I did not see any cases where ios-deploy worked while we were not able to install the app using appium-ios-device lib

**Comment by KazuCocoa** [comment]
70mb ipa, which is 200mb+ `app`

this pr:
```
2024-04-26 06:22:48:442 - [XCUITestDriver@2bdf (3020466d)] App 'com.kazucocoa.sample' is not installed yet or it has an offload and cannot be detected, which might keep the local data.
2024-04-26 06:22:48:442 - [debug] [XCUITestDriver@2bdf (3020466d)] Installing '/Users/kazu/Desktop/sample.ipa' on the device with UUID '242bfcc998fd156df0bd0ad1dde8ab8e0a032114'...
2024-04-26 06:22:54:884 - [debug] [XCUITestDriver@2bdf (3020466d)] A new application installation is going to be performed. Will timeout in 474158.760375 ms
```



default:
```
[debug] [XCUITestDriver@8235 (c10fbeef)] Installing '/var/folders/xt/hbb7vf_d2_g6ks5b765c890c0000gn/T/2024325-656-guiwmj.x8x4s/sample.app' on the device with UUID '242bfcc998fd156df0bd0ad1dde8ab8e0a032114'...
[debug] [XCUITestDriver@8235 (c10fbeef)] Using 'serial' app deployment strategy. You could change it by providing another value to the 'appInstallStrategy' capability
[debug] [XCUITest] Getting bundle ID from app '/var/folders/xt/hbb7vf_d2_g6ks5b765c890c0000gn/T/2024325-656-guiwmj.x8x4s/sample.app': 'com.kazucocoa.sample'
[debug] [XCUITest] Successfully scanned the tree structure of '/var/folders/xt/hbb7vf_d2_g6ks5b765c890c0000gn/T/2024325-656-guiwmj.x8x4s/sample.app'
[debug] [XCUITest] Got 303 folders and 742 files to push
[debug] [XCUITest] Successfully created the remote folder structure (304 items)
[debug] [XCUITest] Proceeding to serial files push
[debug] [XCUITest] Successfully pushed 303 folders and 742 files within 21292ms
```


parallel:
```
[debug] [XCUITestDriver@a904 (d1b51063)] Installing '/var/folders/xt/hbb7vf_d2_g6ks5b765c890c0000gn/T/2024325-527-4st004.jyqlt/sample.app' on the device with UUID '242bfcc998fd156df0bd0ad1dde8ab8e0a032114'...
[debug] [XCUITestDriver@a904 (d1b51063)] Using 'parallel' app deployment strategy. You could change it by providing another value to the 'appInstallStrategy' capability
[debug] [XCUITest] Getting bundle ID from app '/var/folders/xt/hbb7vf_d2_g6ks5b765c890c0000gn/T/2024325-527-4st004.jyqlt/sample.app': 'com.kazucocoa.samole'
[debug] [XCUITest] Successfully scanned the tree structure of '/var/folders/xt/hbb7vf_d2_g6ks5b765c890c0000gn/T/2024325-527-4st004.jyqlt/sample.app'
[debug] [XCUITest] Got 303 folders and 742 files to push
[debug] [XCUITest] Successfully created the remote folder structure (304 items)
[debug] [XCUITest] Proceeding to parallel files push (max 8 writers)
[debug] [XCUITest] Successfully pushed 303 folders and 742 files within 12770ms
```

File: lib/real-device.js [hunk 1]
```diff
@@ -71,82 +69,50 @@ export class RealDevice {
...
-        await pushFolder(afcService, app, bundlePathOnPhone, {
-          enableParallelPush,
-          timeoutMs: timeout,
+    const afcService = await services.startAfcService(this.udid);
+    try {
+      let bundlePathOnPhone;
+      if ((await fs.stat(appPath)).isFile()) {
+        // https://github.com/doronz88/pymobiledevice3/blob/6ff5001f5776e03b610363254e82d7fbcad4ef5f/pymobiledevice3/services/installation_proxy.py#L75
+        bundlePathOnPhone = `/${path.basename(appPath)}`;
+        await pushFile(afcService, appPath, bundlePathOnPhone, {
```
Comment (KazuCocoa) [inline lib/real-device.js:87]: not this pr change itself, but can we add log like `Successfully pushed 303 folders and 742 files within 12807ms` in pushFolder, how long did the push file take, as debug log info? Such log helped in the past

**Review by KazuCocoa — COMMENTED** [review]
Not so large file, but I have tested improvements in a real device (over USB) in some file format, zipped ipa, ipa, zip, app. The behavior itself seems good for now

File: lib/real-device.js (hunk 1 above)
Comment (mykola-mokhnach) [reply lib/real-device.js:87]: added this now

**Comment by mykola-mokhnach** [comment]
> 2024-04-26 06:22:48:442 - [XCUITestDriver@2bdf (3020466d)] App 'com.kazucocoa.sample' is not installed yet or it has an offload and cannot be detected, which might keep the local data.
> 2024-04-26 06:22:48:442 - [debug] [XCUITestDriver@2bdf (3020466d)] Installing '/Users/kazu/Desktop/sample.ipa' on the device with UUID '242bfcc998fd156df0bd0ad1dde8ab8e0a032114'...
> 2024-04-26 06:22:54:884 - [debug] [XCUITestDriver@2bdf (3020466d)] A new application installation is going to be performed. Will timeout in 474158.760375 ms

So this means we took 6s to push the file? Seems a bit slow for 70MB. If we assume the average USB2 transfer limit is 30MB then it should only take 2-3s

**Comment by mykola-mokhnach** [comment]
Could you also please check the application upgrade path on real devices @KazuCocoa ?
The scenario is:
- Install older .ipa via capabilities
- Using installApp extension install a newer version of the same .ipa

**Comment by KazuCocoa** [comment]
Yea, tested the same file with the log
```
2024-04-26 07:02:31:423 - [debug] [XCUITest] Successfully pushed the file payload (75.30 MB) to the remote location '/sample.ipa' in 5918ms
```

**Comment by KazuCocoa** [comment]
Yea, I did that. `appium:enforceAppInstall` combination was also

File: lib/app-utils.js [hunk 2]
```diff
@@ -504,16 +600,24 @@ async function unzipApp(appPathOrZipStream, depth = 0) {
...
+ *
+ * @this {XCUITestDriver}
  * @param {import('@appium/types').DownloadAppOptions} opts
  * @returns {Promise<string>}
  */
-export async function onDownloadApp({stream}) {
-  return await unzipApp.bind(this)(stream);
+export async function onDownloadApp({stream, headers}) {
+  return this.isRealDevice()
+    ? await downloadIpa.bind(this)(stream, headers)
```
Comment (KazuCocoa) [inline lib/app-utils.js:615]: Should here return string of promise instead of IpaInfo?

Comment (mykola-mokhnach) [reply lib/app-utils.js:615]: good catch, thanks. I've changed the return type of downloadIpa helper
weird that typescript did not catch that

**Review by KazuCocoa — APPROVED** [review]
lgtm

**Pull request merged**